### PR TITLE
Fix contour plot of matplotlib

### PR DESCRIPTION
**Pull request by fusawa-yugo**
## Motivation

Outputs of coutour plot are different between plotly and matplotlib. 
Visualization of matplotlib is not working properly.

スクリーンショット 2024-12-27 17 27 32

- how to reproduce
Use these and visualize through `optuna.visualization.matplotlib.plot_contour`
https://github.com/optuna/optuna-visualization-regression-tests/blob/6b98f408f82a613fc5271cfb12308649e3d55211/studies.py#L45-L53
https://github.com/optuna/optuna-visualization-regression-tests/blob/6b98f408f82a613fc5271cfb12308649e3d55211/studies.py#L24-L29

In this case,  if `category` is `bar`, `x1` is missing as shown below.
```
Params: {'category': 'bar', 'x2': -6.85817929203591}
Params: {'category': 'foo', 'x1': 9.955852432587085}
Params: {'category': 'foo', 'x1': 8.553363427670403}
```
Now, when plotting a gragh related to `x1`, `bar` is dropped from `category` axis.
This is the same for `foo` and `x1` .

## Description of the changes

I solved this by including all values on the axis with a categorical distribution, regardless of whether the value on the other axis is None.

contour_mat_4 2dev

### CODE REVIEW

File: optuna/visualization/matplotlib/_contour.py [hunk 1]
```diff
@@ -175,7 +175,9 @@ def _calculate_axis_data(
     returned_values: Sequence[int | float]
     if axis.is_cat:
         enc = _LabelEncoder()
-        returned_values = enc.fit_transform(list(map(str, values)))
+        # prepare all values in categorical distribution
```
Comment (toshihikoyanase) [inline optuna/visualization/matplotlib/_contour.py:178]: A nit, but please follow the coding convention: "Block and inline comments are supposed to be sentence(s)."
https://github.com/optuna/optuna/wiki/Coding-Style-Conventions#block-and-inline-comments-are-supposed-to-be-sentences

```suggestion
        # Prepare all values in categorical distributions.
```

**Comment by HideakiImamura** [comment]
@nabenabe0928 Could you review this PR?

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the PR. The change looks good to me. Is it possible to introduce the test cases to validate this change?

**Comment by nabenabe0928** [comment]
The PR description could be better by the following revisions:


### Describe how they look differently

In my eyes, the difference was not that obvious.
If I understand the difference correctly, each plot includes only a part of categories, making the dots only on the left side or the bottom of each plot in matplotlib.

### Claim which figure is for matplotlib

In the first figure, it is not obvious which figure shows the result obtained by matplotlib, so it would be much more friendly with the guide.

### Add a code block for smoother PR reviews

It is not that hard to add a code block, but it significantly makes reviews easier.

<details>
<summary>Demo Code</summary>

```python
import optuna
import matplotlib.pyplot as plt


def objective(trial: optuna.Trial) -> float:
    category = trial.suggest_categorical("category", ["foo", "bar"])
    if category == "foo":
        return (trial.suggest_float("x1", 0, 10) - 2) ** 2
    else:
        return -((trial.suggest_float("x2", -10, 0) + 5) ** 2)


study = optuna.create_study()
study.optimize(objective, n_trials=50)
fig = optuna.visualization.plot_contour(study)
fig.show()
optuna.visualization.matplotlib.plot_contour(study)
plt.show()

```

</details>

File: optuna/visualization/matplotlib/_contour.py [hunk 2]
```diff
@@ -175,7 +175,9 @@ def _calculate_axis_data(
...
+        enc.fit(list(map(str, filter(lambda value: value is not None, axis.values))))
```
Comment (nabenabe0928) [inline optuna/visualization/matplotlib/_contour.py:179]: Please note the following comment as well:
- https://github.com/optuna/optuna/pull/5892/files#r1898380046

```suggestion
        # Fit LabelEncoder with all the categories in categorical distribution.
        enc.fit(list(map(str, filter(lambda value: value is not None, axis.values))))
        # Then transform the values using the fitted label encoder.
        # Note that `values` may not include all the categories, so we use `axis.values` for fitting.
```

**Comment by nabenabe0928** [comment]
Seems the interpolation algorithm is a bit different from plotly, but that is not a range of this PR, so I will leave it as is.

**Comment by fusawa-yugo** [comment]
I fixed comments.
Could you please review it?

**Review by nabenabe0928 — APPROVED** [review]
LGTM!

**Comment by HideakiImamura** [comment]
@fusawa-yugo Could you add a unittest for this change?

**Comment by fusawa-yugo** [comment]
@HideakiImamura 
I discussed with @c-bata, and I think that translating the expected inputs and outputs of this private function into unit tests is not straightforward in this case. 
Therefore, I feel if it might be reasonable to omit adding unit tests this time.

**Review by HideakiImamura — APPROVED** [review]
Thanks for the discussion. I agree with you. It would be great to work on the test of matplotlib contour tests as a follow-up.
